### ISSUE TITLE
Too small secure buffer size

### ISSUE THREAD

**Issue by Janrupf** [dart-lang/sdk#46476]
# Overview
Secure buffers seem to be hardcoded to 8192 bytes, resulting in slow download speeds. It would be desirable to have the size configurable at runtime.

## Where
https://github.com/dart-lang/sdk/blob/82ea8416ed2a22e40c5c67252cb37bebec3a57e8/sdk/lib/_internal/vm/bin/secure_socket_patch.dart#L73-L76

These 2 sizes are then used to create buffers for SSL operations (see this function):
https://github.com/dart-lang/sdk/blob/master/runtime/bin/secure_socket_filter.cc#L367

## Whats the problem
The buffer sizes limit download speeds. Downloading an 16MB file takes over 20 seconds(!).

People seem to have experienced this in the wild already quite a lot:
https://stackoverflow.com/a/65656847

## Affected platforms
As far as I can tell, everything which runs the DartVM. For flutter this means probably only desktop platforms, but this is a Dart issue anyway.

**Comment by julemand101** [comment]
Do you have an example of some Dart code which shows the problem? I recently helped some guy with some download code and could easily download 800 mb in no time with the code I ended up with, so I don't think your issue here is a general issue. Also, which platforms do you have the problem on?

**Comment by Janrupf** [comment]
The platform I have the problem on is Linux. Basically any secure (https in this context) download using the dart:io HTTP client shows the problem for me.

**Comment by julemand101** [comment]
Here is an example I have written:
```dart
import 'dart:io';

Future<void> main() async {
  final stopWatch = Stopwatch()..start();
  final httpClient = HttpClient();
  final request = await httpClient.getUrl(Uri.parse(
      'https://mirror.one.com/archlinux/iso/2021.06.01/archlinux-2021.06.01-x86_64.iso'));

  final response = await request.close();
  await File('/tmp/archlinux-2021.06.01-x86_64.iso')
      .openWrite()
      .addStream(response);

  stopWatch.stop();
  print('${stopWatch.elapsed}');
}
```

Running on my Arch Linux server, with 100 Mbps download connection, I get the following outout:
```
[julemand101@beta tmp]$ dart test.dart 
0:01:09.400919
[julemand101@beta tmp]$ ls -lsh archlinux-2021.06.01-x86_64.iso 
775M -rw-r--r-- 1 julemand101 julemand101 775M Jun 26 22:31 archlinux-2021.06.01-x86_64.iso
```
Can you show an example I can try run just to see if I can reproduce your problem?

**Comment by Janrupf** [comment]
Alright... it turns out to actually be a problem with my test setup. Well, the more you know.

The issue was the way I was buffering the data, which turned out to be incredibly slow. I'm still quite confused about why I receive the data in 8191 bytes followed by 2 byte batches, which lead to my initial though of the problem being the secure buffer size. 

Thanks for hinting me at the correct implementation, I guess I should have analyzed the issue in a different way.

Closing, not a bug but my fault

**Comment by Janrupf** [comment]
I'll give a quick though on what I think is happening:
If I have a `LinearProgressIndicator` visible on the application, the download time extends from 1 second to **26** seconds. I did not start the download in a new isolate, so my guess is that the UI update task for the progress indicator takes up the time and slows down the download callback (I'm buffering data into a List).

Starting the download in a new isolate randomly hangs it... but this doesn't fit this issue anymore

**Comment by julemand101** [comment]
@Janrupf 
Your problem is properly that you are firing an redraw event on the `LinearProgressIndicator` every time you get an event on the network stream which happens to be in very small sizes indeed. I don't know your code but you should add some logic so you are only sends an update to the progress bar when e.g. 1-5% of the total download has been done.

**Comment by quambene** [comment]
@Janrupf I believe to have a similar issue with RefreshIndicator (api request slowing down if response body > 100 kb for package "http/http.dart"). Did you solve your issue? Can you share your solution?

**Comment by Janrupf** [comment]
TL;DR: Perform the download in a separate isolate, make sure its main function does not terminate before the download finishes

My solution was simply to not run the HTTP request on the UI isolate. As mentioned before, the HTTP client shares the event loop with UI updates. This means that if running on the UI isolate, and some other element is active on it (an animating progress bar for example, the HTTP client does not get enough processing time to perform a fast download.

The solution is to perform the download on another isolate (where the HTTP client has its own event loop, don't perform other tasks on that isolate or you will have issues again). Also make sure to hang the Isolate main function until the download completes, else it will terminate the event loop before the download finishes (thus effectively cancelling the download).

**Comment by quambene** [comment]
Thanks for your explanation. It seems that my issue is independent from this. There is a huge performance drop from ~100 milliseconds to ~ 5 seconds if request body gets greater than 100 kb. That is, even for 101 kb the request takes ~5 seconds. Will check if this can also be reproduced in the dart:io package. I was using the http/http.dart package.
